Mathlib/Geometry/Group/Marking.lean

```diff
@@ -0,0 +1,186 @@
+
+@[to_additive]
+instance MarkedGroup.instMulAction [MulAction G α] : MulAction (MarkedGroup m) α := ‹MulAction G α›
+
+@[to_additive (attr := simp)]
+lemma toMarkedGroup_smul (g : G) (x : α) [SMul G α] :
+    (toMarkedGroup g : MarkedGroup m) • x = g • x := rfl
+
+@[to_additive (attr := simp)]
+lemma ofMarkedGroup_smul (g : MarkedGroup m) (x : α) [SMul G α] : ofMarkedGroup g • x = g • x := rfl
+
+@[to_additive]
+private lemma mul_aux [DecidableEq S] (x : MarkedGroup m) :
+    ∃ (n : _) (l : FreeGroup S), toMarkedGroup (m l) = x ∧ l.toWord.length ≤ n := by
+  classical
+  obtain ⟨l, rfl⟩ := m.coe_surjective x
+  exact ⟨_, _, rfl, le_rfl⟩
+
+@[to_additive]
+private lemma mul_aux' [DecidableEq S] (x : MarkedGroup m) :
+    ∃ (n : _) (l : FreeGroup S), toMarkedGroup (m l) = x ∧ l.toWord.length = n := by
+  classical
+  obtain ⟨l, rfl⟩ := m.coe_surjective x
+  exact ⟨_, _, rfl, rfl⟩
+
+@[to_additive]
+private lemma find_mul_aux [DecidableEq S] (x : MarkedGroup m)
+    [DecidablePred fun n ↦ ∃ l, toMarkedGroup (m l) = x ∧ l.toWord.length ≤ n]
+    [DecidablePred fun n ↦ ∃ l, toMarkedGroup (m l) = x ∧ l.toWord.length = n] :
+    Nat.find (mul_aux x) = Nat.find (mul_aux' x) := by
+  classical
+  exact _root_.le_antisymm (Nat.find_mono fun n => Exists.imp fun l => And.imp_right le_of_eq) <|
+    (Nat.le_find_iff _ _).2 fun k hk ⟨l, hl, hlk⟩ => (Nat.lt_find_iff _ _).1 hk _ hlk ⟨l, hl, rfl⟩
+
+@[to_additive]
+noncomputable instance : NormedGroup (MarkedGroup m) :=
+  GroupNorm.toNormedGroup
+    { toFun := fun x => by classical exact (Nat.find (mul_aux x)).cast
+      map_one' := by
+        classical
+        exact cast_eq_zero.2 <| (find_eq_zero <| mul_aux _).2 ⟨1, by simp_rw [map_one], le_rfl⟩
+      mul_le' := fun x y => by
+        classical
+        dsimp
+        norm_cast
+        obtain ⟨a, rfl, ha⟩ := Nat.find_spec (mul_aux x)
+        obtain ⟨b, rfl, hb⟩ := Nat.find_spec (mul_aux y)
+        refine find_le ⟨a * b, by simp, (a.toWord_mul_sublist _).length_le.trans ?_⟩
+        rw [length_append]
+        gcongr
+      inv' := by
+        classical
+        suffices ∀ {x : MarkedGroup m}, Nat.find (mul_aux x⁻¹) ≤ Nat.find (mul_aux x) by
+          dsimp
+          refine fun _ => congr_arg Nat.cast (this.antisymm ?_)
+          convert this
+          simp_rw [inv_inv]
+        rintro x
+        refine find_mono fun nI => ?_
+        rintro ⟨l, hl, h⟩
+        exact ⟨l⁻¹, by simp [hl], h.trans_eq' <| by simp⟩
+      eq_one_of_map_eq_zero' := fun x hx => by
+        classical
+        obtain ⟨l, rfl, hl⟩ := (find_eq_zero <| mul_aux _).1 (cast_eq_zero.1 hx)
+        rw [le_zero_iff, length_eq_zero, ← FreeGroup.toWord_one] at hl
+        rw [FreeGroup.toWord_injective hl, map_one, map_one] }
+
+namespace MarkedGroup
+
+@[to_additive]
+lemma norm_def [DecidableEq S] (x : MarkedGroup m)
+    [DecidablePred fun n ↦ ∃ l, toMarkedGroup (m l) = x ∧ l.toWord.length = n] :
+    ‖x‖ = Nat.find (mul_aux' x) := by
```

Comment:
It's generally not a great idea to have a lemma of this form, where the definition depends on a private lemma. You should instead give properties of the norm which can be useful without referencing private lemmas. For instance, the lemma that there exists a word in the free group with length n which corresponds to x is tricky to prove from the public API. This, and related lemmas (the norm is the smallest n with that property) should be included as part of the norm API.